libcxx/src/locale.cpp

```diff
@@ -538,16 +549,40 @@ locale::__imp::use_facet(long id) const
 
 // locale
 
+std::atomic<locale::__imp*> locale::__imp::classic_;
```

Comment:
I think we can do this whole thing differently. First, I suggest a refactor in https://github.com/llvm/llvm-project/pull/72581. Then, I think we can implement this patch as:

```
id locale::__imp::acquire() {
    if (this != &locale::classic().__locale_)
        __add_shared();
}
```

and correspondingly for `release()`. And we wouldn't need to maintain or store a separate `atomic<__imp*>`. The difference is that with the currently-proposed patch, we pay for one relaxed atomic load when we call `release()` or `acquire()`. With my proposed approach above, we instead pay for a call to `__cxa_acquire` when we enter `locale::classic()`. I'd be curious to see how the two approaches compare performance-wise, but I suspect that any difference might be in the noise.

Reply:
I've considered this and specifically did with the additional variable for performance reasons.
There is no cost for atomic relaxed loads. On any sane architecture it's a normal load.
While the access to the local static object with dynamic initialization will incur:
1. An atomic acquire load (which is costlier and arm64).
2. A branch to handle the lazy dynamic initialization.
3. The indirection to load imp pointer from the classic locale object, so an additional cache line touched and a potential cache miss/eviction.